### PR TITLE
feat: support aten._cdist_forward converter

### PR DESCRIPTION
**Pull request by chohk88**
# Description

New feature to support aten._cdist_forward converter converter. I have added test cases ensuring compatibility with both matching and broadcasting input shapes.

Fixes # ([issue](https://github.com/pytorch/TensorRT/issues/2725))

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 1]
```diff
@@ -2133,6 +2133,27 @@ def aten_ops_linear(
     )
 
 
+@dynamo_tensorrt_converter(torch.ops.aten._cdist_forward.default)
+@dynamo_tensorrt_converter(torch.ops.aten._cdist_forward)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:2137]: Typically, `torch.ops.aten._cdist_forward` is equal to `torch.ops.aten._cdist_forward.default`. (except in some rare cases) If deleting `torch.ops.aten._cdist_forward` works, then please delete it.

File: py/torch_tensorrt/dynamo/conversion/impl/linear.py
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/linear.py:1]: `cdist` looks similar to `pdist` so they could be put together under `normalization` folder.

File: py/torch_tensorrt/dynamo/conversion/impl/linear.py [hunk 2]
```diff
@@ -52,3 +56,107 @@ def linear(
...
+
+def cdist_forward(
+    ctx: ConversionContext,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    x1: TRTTensor,
+    x2: TRTTensor,
+    p: float,
+    compute_mode: Optional[int] = None,
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/linear.py:69]: It seems that `compute_mode` is not used below. Is there any difficulty for this? Ideally, we should mimic all pytorch's behaviors (i.e., receiving the same args and then returning the same outputs as pytorch, based on the [schema](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L4406C60-L4406C72)). 
If `compute_mode` cannot be supported for some reasons, we need to use `capability_validator` to make sure this kind of cases will not be converted, instead, falling back to pytorch.

File: tests/py/dynamo/conversion/test_cdist_aten.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            ((15, 10, 5), 2.0),
+            ((15, 10, 5), 2.99),
+            ((15, 10, 5), float("inf")),
+        ]
+    )
+    def test_cdist_float(self, shape, p):
+        class Cdist(nn.Module):
+            def forward(self, x1, x2):
+                print("x1 : ", x1)
+                print("x2 : ", x2)
```
Comment (zewenli98) [inline tests/py/dynamo/conversion/test_cdist_aten.py:27]: The print should be removed.

File: tests/py/dynamo/conversion/test_cdist_aten.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+                return torch.ops.aten._cdist_forward.default(x1, x2, p, None)
```
Comment (zewenli98) [inline tests/py/dynamo/conversion/test_cdist_aten.py:29]: Different `compute_mode`s other than `None` should be tested.

File: tests/py/dynamo/conversion/test_cdist_aten.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+        inputs = [torch.randn(shape), torch.randn(shape)]
```
Comment (zewenli98) [inline tests/py/dynamo/conversion/test_cdist_aten.py:31]: Per the pytorch [doc](https://pytorch.org/docs/stable/generated/torch.cdist.html), it seems that `x1` and `x2` could be different shape. If then, can you add more tests for different shapes?

File: tests/py/dynamo/conversion/test_cdist_aten.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            ((2, 2, 4, 5), (2, 3, 5), 2.99),
+            ((2, 2, 4, 5), (2, 3, 5), float("inf")),
+        ]
+    )
+    def test_cdist_broadcast_float(self, shape_1, shape_2, p):
+        class Cdist(nn.Module):
+            def forward(self, x1, x2):
+                return torch.ops.aten._cdist_forward.default(x1, x2, p, None)
+
+        inputs = [torch.randn(shape_1), torch.randn(shape_2)]
```
Comment (zewenli98) [inline tests/py/dynamo/conversion/test_cdist_aten.py:56]: Same suggestions as above

**Review by zewenli98 — CHANGES_REQUESTED** [review]

File: py/torch_tensorrt/dynamo/conversion/impl/linear.py
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/linear.py:1]: I have moved the `_cdist_forward` converter to the `normalization` folder.

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 1 above)
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: I have deleted `@dynamo_tensorrt_converter(torch.ops.aten._cdist_forward)` and verified that it works.

File: py/torch_tensorrt/dynamo/conversion/impl/linear.py (hunk 2 above)
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/linear.py]: `compute_mode` does not influence the output values. It is designed for optimizing computational efficiency in specific situations when calculating the Euclidean distance (p=2) for large-sized inputs. 

I have modified to clarify:
1) A note has been added that `compute_mode` doesn't alter the computational path in our implementation.
2) A default value of 0 is used for `compute_mode` when it's unspecified, maintaining behavior consistency with Pytorch.
3) I have included a warning for situations where an optimized path for p=2 might be expected but isn't utilized due to our unified approach (element-wise).

File: tests/py/dynamo/conversion/test_cdist_aten.py (hunk 3 above)
Comment (chohk88) [reply tests/py/dynamo/conversion/test_cdist_aten.py]: Thanks! It seems I overlooked removing an unnecessary debug code

File: tests/py/dynamo/conversion/test_cdist_aten.py (hunk 5 above)
Comment (chohk88) [reply tests/py/dynamo/conversion/test_cdist_aten.py:28]: In the `test_cdist_float_broadcast_and_diff_shape` function, I have added tests for scenarios where the two inputs require broadcasting, as well as for cases where the shapes of the inputs are completely distinct. 

Additionally, I have adjusted the values of input shapes to enable testing with a wider variety of inputs.

File: tests/py/dynamo/conversion/test_cdist_aten.py (hunk 6 above)
Comment (chohk88) [reply tests/py/dynamo/conversion/test_cdist_aten.py]: Thank you! 

In this function, I was already testing with different shapes. To clarify further, I've added `names` to the parameters of the test cases and distinguished between cases `where broadcasting is required with different input shapes` and cases `where the input shapes are different and broadcasting is not needed`.

File: py/torch_tensorrt/dynamo/conversion/impl/linear.py (hunk 2 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/linear.py]: Thanks for the explanation! I'm still thinking if using `matmul` will be faster for the previous two scenarios? Considering that we have out-of-the-box [matmul](https://github.com/pytorch/TensorRT/blob/main/py/torch_tensorrt/dynamo/conversion/impl/matmul.py#L13), if possible, could you compare their time costs?

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 7]
```diff
@@ -2133,6 +2133,26 @@ def aten_ops_linear(
...
+) -> Union[TRTTensor, Sequence[TRTTensor]]:
+    return impl.normalization.cdist_forward(
+        ctx,
+        target,
+        SourceIR.ATEN,
+        name,
+        x1=args[0],
+        x2=args[1],
+        p=args[2],
+        compute_mode=args[3],
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:2152]: Based on the [schema](https://github.com/pytorch/pytorch/blob/7b11fb4695c61e04c7032e039f0a9f02636e0176/aten/src/ATen/native/native_functions.yaml#L4415) having `int?` as the type for `compute_mode`, this should be `compute_mode = args_bounds_check(args, 2, None)`

File: py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py [hunk 8]
```diff
@@ -508,3 +511,119 @@ def pdist(
...
+
+def cdist_forward(
+    ctx: ConversionContext,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    x1: TRTTensor,
+    x2: TRTTensor,
+    p: float,
+    compute_mode: int,
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py:524]: Switch to `Optional[int]`, since `None` is a valid choice

File: py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py [hunk 9]
```diff
@@ -508,3 +511,119 @@ def pdist(
...
+            target,
+            source_ir,
+            f"{name}_max",
+            abs_val,
+            dim=-1,
+            keepdim=False,
+            return_indices=False,
+        )
+    else:
+        raise NotImplementedError(f"Currently, p={p} is not implemented.")
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py:628]: This can be removed, since it seems you have covered all of the valid intervals by Torch's definition of `cdist`, in $\[0,\infty\]$, as per [this](https://pytorch.org/docs/stable/generated/torch.cdist.html)

**Review by gs-olive — COMMENTED** [review]
Overall looks good - added a few suggestions

File: py/torch_tensorrt/dynamo/conversion/impl/linear.py (hunk 2 above)
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/linear.py]: Thank you for your valuable comment. In this update, I've implemented a matrix multiplication-based approach to compute the distance for `p=2`. During my tests on my local PC with inputs `x1=(150,100,50,50)` and `x2=(150,100,30,50)`, I observed that the TRT run time improved by approximately `6.3 times`. Here are the relevant logs from those tests:

```
INFO:harness:FX graph= graph():
    %x1 : [num_users=1] = placeholder[target=x1]
    %x2 : [num_users=1] = placeholder[target=x2]
    %_cdist_forward_default : [num_users=1] = call_function[target=torch.ops.aten._cdist_forward.default](args = (%x1, %x2, 2, 0), kwargs = {})
    return _cdist_forward_default
===============based on matmul : ==================
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.007999
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:255: DeprecationWarning: Use build_serialized_network instead.
  engine = self.builder.build_engine(self.ctx.net, builder_config)
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:33.710717
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 244801024 bytes of Memory
INFO:harness:Interpreter run time(s): 33.719337898997765
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:63: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.input_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:66: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.output_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:88: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:93: DeprecationWarning: Use get_tensor_shape instead.
  tuple(self.engine.get_binding_shape(idx))
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:98: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:247: DeprecationWarning: Use set_input_shape instead.
  self.context.set_binding_shape(
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:262: DeprecationWarning: Use get_tensor_shape instead.
  shape = tuple(self.context.get_binding_shape(idx))
INFO:harness:TRT run time(s)= 0.0026535038948059084
.INFO:harness:FX graph= graph():
    %x1 : [num_users=1] = placeholder[target=x1]
    %x2 : [num_users=1] = placeholder[target=x2]
    %_cdist_forward_default : [num_users=1] = call_function[target=torch.ops.aten._cdist_forward.default](args = (%x1, %x2, 2, 1), kwargs = {})
    return _cdist_forward_default
===============based on elementwise pow for diff : ==================
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.003120
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:255: DeprecationWarning: Use build_serialized_network instead.
  engine = self.builder.build_engine(self.ctx.net, builder_config)
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:43.902498
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 4590000640 bytes of Memory
INFO:harness:Interpreter run time(s): 43.90606521600421
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:63: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.input_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:66: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.output_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:88: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:93: DeprecationWarning: Use get_tensor_shape instead.
  tuple(self.engine.get_binding_shape(idx))
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:98: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:247: DeprecationWarning: Use set_input_shape instead.
  self.context.set_binding_shape(
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:262: DeprecationWarning: Use get_tensor_shape instead.
  shape = tuple(self.context.get_binding_shape(idx))
INFO:harness:TRT run time(s)= 0.016830976486206056
.
----------------------------------------------------------------------
```

File: py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py (hunk 8 above)
Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py]: Thank you for your comment! I've made the requested changes.

File: py/torch_tensorrt/dynamo/conversion/impl/linear.py (hunk 2 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/linear.py]: It seems `matmul` runs faster in both compilation stage and running stage, and save a lot of memory.

Comment (chohk88) [reply py/torch_tensorrt/dynamo/conversion/impl/linear.py]: You're right. While always using matrix multiplication might seem more efficient considering TensorRT's capabilities, I've aligned the implementation with the compute mode of the ATen operations for consistency. Thank you for pointing this out!

**Comment by zewenli98** [comment]
@chohk88 PR #2819 was merged to the main. Can you rebase and check if the error in CI is fixed?

**Review by zewenli98 — CHANGES_REQUESTED** [review]
@chohk88 It seems the tests have too large dataset, which caused OOM error in CI. My advice is to turn them down but make sure every condition path would be covered.

**Comment by chohk88** [comment]
@zewenli98 Thank you for your comments! The issue with CI/CD failures due to the dataset size has been resolved. However, while the Windows wheel build has completed successfully, the Linux wheel build is still failing.

**Comment by zewenli98** [comment]
> @zewenli98 Thank you for your comments! The issue with CI/CD failures due to the dataset size has been resolved. However, while the Windows wheel build has completed successfully, the Linux wheel build is still failing.

It seems to work now. LGTM.
